field/gf251/element.go

```diff
@@ -0,0 +1,160 @@
+// Copyright 2025 Consensys Software Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Code generated by go-corset DO NOT EDIT
+
+package gf251
+
+import (
+	"cmp"
+	"math/bits"
+	"strconv"
+)
+
+// Element of a prime order field, represented in Montgomery form to speed up multiplications.
+type Element [1]uint32 // defined as an array to prevent mistaken use of arithmetic operators, or naive assignments.
+
+const (
+	r                 = 1 << 32 // register size
+	modulus           = 251
+	rSq               = 69         // r² (mod m)
+	negModulusInvModR = 1711142349 // -modulus⁻¹ (mod r), used for Montgomery reduction
+)
+
+// Add x + y
+func (x Element) Add(y Element) Element {
+	res := x[0] + y[0]
+
+	if reduced, borrow := bits.Sub32(res, modulus, 0); borrow == 0 {
+		res = reduced
+	}
+
+	return Element{res}
+}
+
+// Sub x - y
+func (x Element) Sub(y Element) Element {
+	res, borrow := bits.Sub32(x[0], y[0], 0)
+	if borrow != 0 {
+		res += modulus
+	}
+
+	return Element{res}
+}
+
+// montgomeryReduce x -> x.R⁻¹ (mod m)
+func montgomeryReduce(x uint64) Element {
+	// textbook Montgomery reduction
+	m := (x * uint64(negModulusInvModR)) % r // m = x * (-modulus⁻¹) (mod r)
+
+	res := uint32((x + m*modulus) / r)
```

Comment:
I seem confused about this.  Are you just working on the assumption that the compiler will turn `/ r` into `>> 5` ?

Reply:
Yes I'm assuming division and modulus operations against a constant power of two is optimized into shifts and ands.